packages/assets-controllers/src/NftDetectionController.test.ts

```diff
@@ -870,40 +524,142 @@ describe('NftDetectionController', () => {
-        isCurrentlyOwned: true,
-        creator: {
-          user: { username: '' },
-          profile_img_url: '',
-          address: '',
+        options: {
+          addNft: mockAddNft,
+          getNftApi: jest
+            .fn()
+            .mockImplementation(
+              ({
+                contractAddress,
+                tokenId,
+              }: {
+                contractAddress: string;
+                tokenId: string;
+              }) =>
+                `${OPENSEA_PROXY_URL}/chain/ethereum/contract/${contractAddress}/nfts/${tokenId}`,
+            ),
         },
       },
-    ]);
+      async ({ controller, triggerPreferencesStateChange }) => {
+        triggerPreferencesStateChange({
+          ...getDefaultPreferencesState(),
+          selectedAddress,
+          useNftDetection: true,
+        });
+        // Wait for detect call triggered by preferences state change to settle
+        await advanceTime({
+          clock,
+          duration: 1,
+        });
+        mockAddNft.mockReset();
+
+        await controller.detectNfts();
+
+        expect(mockAddNft).toHaveBeenCalledWith(nftContract, nftId, {
+          nftMetadata: {
+            imageOriginal: 'https://example.com/image.gif',
+            name: undefined,
+            standard: 'ERC721',
+            creator: {
+              user: { username: '' },
+              profile_img_url: '',
+              address: '',
+            },
+          },
+          userAddress: selectedAddress,
+          source: Source.Detected,
+          networkClientId: undefined,
+        });
+      },
+    );
   });
 });
+
+type WithControllerCallback<ReturnValue> = ({
+  controller,
+}: {
+  controller: NftDetectionController;
+  triggerNftStateChange: (state: NftState) => void;
+  triggerPreferencesStateChange: (state: PreferencesState) => void;
+}) => Promise<ReturnValue> | ReturnValue;
+
+type WithControllerOptions = {
+  options?: Partial<ConstructorParameters<typeof NftDetectionController>[0]>;
+  config?: Partial<NftDetectionConfig>;
+};
+
+type WithControllerArgs<ReturnValue> =
+  | [WithControllerCallback<ReturnValue>]
+  | [WithControllerOptions, WithControllerCallback<ReturnValue>];
+
+/**
+ * Builds a controller based on the given options, and calls the given function
+ * with that controller.
+ *
+ * @param args - Either a function, or an options bag + a function. The options
+ * bag accepts controller options and config; the function
+ * will be called with the built controller.
+ * @returns Whatever the callback returns.
+ */
+async function withController<ReturnValue>(
+  ...args: WithControllerArgs<ReturnValue>
+): Promise<ReturnValue> {
+  const [{ ...rest }, fn] = args.length === 2 ? args : [{}, args[0]];
+  const { options, config } = rest;
+
+  const getNetworkClientById = jest.fn().mockImplementation(() => {
+    return {
+      configuration: {
+        chainId: ChainId.mainnet,
+      },
+      provider: jest.fn(),
+      blockTracker: jest.fn(),
```

Comment:
Hmm, these aren't functions. Could we create a FakeProvider and then use to construct a block tracker? If that's too much trouble, that's fine, but just a thought.

Reply:
Good point, though I'm inclined to leave this alone for now because this mock is pre-existing. This PR just moves it into a helper function.

Reply:
Okay, no problem.